microbench/src/main/java/io/netty/microbench/concurrent/DefaultPromiseBenchmark.java

```diff
@@ -0,0 +1,162 @@
+    private EventExecutor executor;
+    private TestGenericFutureListener[] listeners;
+    private TestGenericFutureListener[] lastListeners;
+    private static AtomicLong doneCount = new AtomicLong();
+    private static CountDownLatch testLatch;
+    private final Runnable toEventLoopRunnable = new Runnable() {
+        @Override
+        public void run() {
+            startOutEventLoop();
+        }
+    };
+
+    @Param({ "100000", "1000000" })
+    public int numPromises;
+
+    @Param({ "1", "4" })
+    public int numListeners;
+
+    @Param({ "0", "4" })
+    public int lateListeners;
+
+    @AuxCounters
+    @State(Scope.Thread)
+    public static class DoneCounter {
+        @Setup(Level.Iteration)
+        public void clean() {
+            doneCount.set(0);
+        }
+
+        public long doneCount() {
+            return doneCount.get();
+        }
+    }
+
+    @Setup(Level.Trial)
+    public void setupTrial() {
+        executorService = Executors.newSingleThreadExecutor();
+        executor = new DefaultEventExecutor(executorService);
+        listeners = new TestGenericFutureListener[numListeners];
+        for (int i = 0; i < listeners.length; ++i) {
+            listeners[i] = new TestGenericFutureListener();
+        }
+        lastListeners = new TestGenericFutureListener[lateListeners];
+        for (int i = 0; i < lastListeners.length; ++i) {
+            lastListeners[i] = new TestGenericFutureListener();
+        }
+    }
+
+    @Setup(Level.Invocation)
+    public void setupInvocation() {
+        testLatch = new CountDownLatch(numPromises * (listeners.length + lastListeners.length));
+    }
+
+    @TearDown(Level.Trial)
+    public void teardownTrial() {
+        final Future<?> f;
+        try {
+            executorService.shutdown();
+        } finally {
+            f = executor.shutdownGracefully(0, 0, TimeUnit.MILLISECONDS);
+        }
+        try {
+            executorService.awaitTermination(5, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+        try {
+            f.get(5, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        } catch (ExecutionException e) {
+            e.printStackTrace();
+        } catch (TimeoutException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Benchmark
+    public void testInEventLoop(DoneCounter doneCounter) throws Exception {
+        startInEventLoop();
+        testLatch.await();
+    }
+
+    @Benchmark
+    public void testOutEventLoop(DoneCounter doneCounter) throws Exception {
+        startOutEventLoop();
+        testLatch.await();
+    }
+
+    private void startInEventLoop() {
+        executor.execute(toEventLoopRunnable);
+    }
+
+    private void startOutEventLoop() {
+        for (int i = 0; i < numPromises; ++i) {
+            Promise<Void> p = new DefaultPromise<Void>(executor);
+            p.addListeners(listeners);
```

Comment:
I had seen your comment, but it didn't really help me understand. Optimizing `addListeners` seems limited benefit, since nobody calls it; there does not seem to be a single usage of the method in Netty (including in a test! It is untested!) and I can't think of any common coding paradigm that would use it. Using `addListeners` would not be a '"typical use case" w.r.t. Promise' as you said in that comment. If instead the purpose was to benchmark listener completion time, then I could understand.

However, that didn't make sense to me either since in the "typical use case" `setSuccess()` is almost always called from the event loop. With `addListener` vs `addListeners`, addListeners is able to run on the event loop as a batch, which is atypical. And since there may be queuing on the event loop adding delay, there may be zero contention. In either case I would expect Java to elide some of the locking, even in the addListener case, since it doesn't seem to hard to realize the calls can share the lock acquisition and release at the end.